Support tool filtering by model

bhavyaus opened the pull request:
For https://github.com/microsoft/vscode/issues/275679

Ensures `ChatRequest.tools` correctly reflects tool availability and enablement state by filtering based on the active model's compatibility.

Key changes:
- Added `supportedModels` field to `IToolDataDto` and passed through main-to-exthost communication
- Updated `ExtHostChatAgents2.getToolsForRequest()` to accept `modelId` parameter and filter tools using `isToolAvailableForModel()`
- Modified tool map to include all model-compatible tools with correct enablement state (defaults to enabled if not explicitly set)
- Updated all `getToolsForRequest()` call sites to pass the active model ID

bhavyaus committed:
Support tool filtering by model

roblourens commented on src/vs/workbench/contrib/chat/common/tools/languageModelToolsContribution.ts:
```diff
@@ -138,6 +139,13 @@ const languageModelToolsExtensionPoint = extensionsRegistry.ExtensionsRegistry.r
 						type: 'string',
 						pattern: '^(?!copilot_|vscode_)'
 					}
+				},
```
Is a static contribution flexible enough? It wouldn't let us control this with a setting/exp, or do more flexible checks like `family.contains('-codex')` that we do in code.

Also, going along with wanting to do other prompt tweaks in a more centralized place, it would be nice to have one place in code where I can determine the prompts and tools enabled for a certain model family.

How do you see this working for websearch, do you register a tool in api that doesn't have a invoke implementation but can show up in UI?

connor4312 replied:
I agree that we probably want more flexibility here. Otherwise every single release of every model will require changes. In the related issue #277472 I brought up the notion of using a `when` clause.

This does not necessarily mean you need to drive it off global context keys (though maybe using `when` would overload it, perhaps use a different name?) as there is a standalone `Parser` in contextkey.ts that returns an expression you can easily evaluate by passing an object with a `getValue` method. So having something specific there that is evaluated when a request is sent with information about the model, family, etc. is doable.

bhavyaus replied:
How about updating `LanguageModelTool`  to allow tools to dynamically specify which models they support?

```
declare module 'vscode' {

	export interface LanguageModelTool<T> {
		/**
		 * Called to check if this tool supports a specific language model. If this method is not implemented,
		 * the tool is assumed to support all models.
		 *
		 * This method allows extensions to dynamically determine which models a tool can work with,
		 * enabling fine-grained control over tool availability based on model capabilities.
		 *
		 * @param modelId The identifier of the language model (e.g., 'gpt-4o', 'claude-3-5-sonnet')
		 * @returns `true` if the tool supports the given model, `false` otherwise
		 */
		supportsModel?(modelId: string): Thenable<boolean>;
	}
}

```

bhavyaus replied:
Was only looking to support this for the builit-in tools we expose. @connor4312 do you anticipate needing this per model activation for MCP tools ?

connor4312 replied:
For MCP there are two scenarios:

1. Tools from MCP servers that are defined in `agents.md` and similar. We're seeing clients start to support inline MCP's in this way and so their tools should only be applied to requests that involve those agent instructions. 
2. Sampling with tools. In these cases tools are purely internal, we get them in core in the MCP sampling request and can register them however we want.

In both cases no particular extension API is needed for the MCP scenarios.

bhavyaus committed:
add supportsModel API to language model tools for dynamic model compatibility

bhavyaus commented on src/vs/workbench/api/common/extHostLanguageModelTools.ts:
```diff
@@ -171,6 +171,28 @@ export class ExtHostLanguageModelTools implements ExtHostLanguageModelToolsShape
+			return undefined;
+		}
+
+		// supportsModel is a proposed API, cast to access it
```
Copilot what?

bhavyaus committed:
refactor: simplify access to supportsModel API in ExtHostLanguageModelTools

bhavyaus committed:
Update invokeTool to support Language

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
chat: wip on model-specific tools

This is a rethinking of the initial API proposed in `languageModelToolSupportsModel.d.ts`.

1. This switches to `models?: LanguageModelChatSelector[];` to control enablement.
   definitely open to switching this out, but I think a synchronously-analyzable
   expression is important to retain the data flows in core without too many races.
2. The extension is able to define a tool at runtime via registerToolDefinition. This
   should let us have entirely service-driven tools from model providers without
   requiring a static definition for each one. We can also have model-specific
   variants of tools without a ton of package.json work for each variant of the tool
   (as initially proposed using `when` clauses)

This then propagates that down into the tools service. Currently I have this as just
compiling to a `when` expression once it reaches the main thread. Then, for the tools
service, it takes an IContextKeyService in cases where tools should be enumerated,
and the chat input sets the model keys in its scoped context key service. This allows
the tools to be filtered correctly in the tool picker.

I initially thought about allowing multiple definitions be registered for the same tool
name/id for model-specific variants of tools but I realized that gets really gnarly and
we already have a `toolReferenceName` that multiple tools can register into.

Todo for tomorrow morning:
- Tools don't make it to the ChatRequest yet, or something, still need to investigate
- Need to make sure tools in prompts/models all work. For a first pass I think we can
  let prompts/modes reference all tools by toolReferenceName.
- Validate that multiple tools actually can safely share a reference name (and do
  some priority ordering?)
- General further validation
- Some unit tests

connor4312 committed:
key selected tools based on reference name

connor4312 committed:
chat: wip on model-specific tools (#287666)

* chat: wip on model-specific tools

This is a rethinking of the initial API proposed in `languageModelToolSupportsModel.d.ts`.

1. This switches to `models?: LanguageModelChatSelector[];` to control enablement.
   definitely open to switching this out, but I think a synchronously-analyzable
   expression is important to retain the data flows in core without too many races.
2. The extension is able to define a tool at runtime via registerToolDefinition. This
   should let us have entirely service-driven tools from model providers without
   requiring a static definition for each one. We can also have model-specific
   variants of tools without a ton of package.json work for each variant of the tool
   (as initially proposed using `when` clauses)

This then propagates that down into the tools service. Currently I have this as just
compiling to a `when` expression once it reaches the main thread. Then, for the tools
service, it takes an IContextKeyService in cases where tools should be enumerated,
and the chat input sets the model keys in its scoped context key service. This allows
the tools to be filtered correctly in the tool picker.

I initially thought about allowing multiple definitions be registered for the same tool
name/id for model-specific variants of tools but I realized that gets really gnarly and
we already have a `toolReferenceName` that multiple tools can register into.

Todo for tomorrow morning:
- Tools don't make it to the ChatRequest yet, or something, still need to investigate
- Need to make sure tools in prompts/models all work. For a first pass I think we can
  let prompts/modes reference all tools by toolReferenceName.
- Validate that multiple tools actually can safely share a reference name (and do
  some priority ordering?)
- General further validation
- Some unit tests

* key selected tools based on reference name

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
fix tests

connor4312 committed:
tests

connor4312 commented:
@aeschli in this PR I have suggested moving from storing the `tool.id` as the key for enablement to instead storing the `toolReferenceName` when possible. This allows two differing tool registrations (e.g. a web search tool from Anthropic and one from OAI) to both share the same name for the user and have their enablement persist together, while having different underlying `id`s, schemas, and implementation.

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
pass the lm directly to the tool service for filtering

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
tools: allow contributing runtime tools to toolsets

connor4312 committed:
Revert "key selected tools based on reference name"

This reverts commit 13a39cc1e2df6b18aa3c7d0ca64b39a6dfbe72eb.

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

aeschli commented:
I like the addition of `toolset` to the tool definition. 
On the [IToolData in core](https://github.com/microsoft/vscode/blob/72bacda9f3493262af0705a4314ffa5ea4ac5a56/src/vs/workbench/contrib/chat/common/tools/languageModelToolsService.ts#L46) I would now add a property `fullToolReferenceName` with derived value `${toolset}/${toolReferenceName ?? displayName`. That should eliminates a lot of complexity on my side and we can remember enabled tools by 'fullToolReferenceName'.

aeschli commented:
Is the main purpose to have specialized tool implementations for a tool? It might be cleaner to model that and have an additional contribution point or provider for tool implementations. A tool implementation has the list of models it supports and points to a tool definition id. It doesn't need to repeat all the properties.
Maybe just the extension needs to deal with the tool implementations

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model

connor4312 committed:
Merge remote-tracking branch 'origin/main' into dev/bhavyau/tools-model
